wire-schema/src/main/java/com/squareup/wire/schema/Linker.java

```diff
@@ -28,117 +28,79 @@
 
 /** Links local field types and option types to the corresponding declarations. */
 final class Linker {
-  private final ImmutableList<ProtoFile> protoFiles;
+  private final Map<String, FileLinker> fileLinkers;
   private final Map<String, Type> protoTypeNames;
+  private final Multimap<String, String> publicImports;
   private final Multimap<String, String> imports;
   private final List<String> errors;
   private final List<Object> contextStack;
 
-  Linker(Iterable<ProtoFile> protoFiles) {
-    this.protoFiles = ImmutableList.copyOf(protoFiles);
+  Linker() {
+    this.fileLinkers = new LinkedHashMap<>();
     this.protoTypeNames = new LinkedHashMap<>();
+    this.publicImports = LinkedHashMultimap.create();
     this.imports = LinkedHashMultimap.create();
     this.contextStack = Collections.emptyList();
     this.errors = new ArrayList<>();
   }
 
   private Linker(Linker enclosing, Object additionalContext) {
-    this.protoFiles = enclosing.protoFiles;
+    this.fileLinkers = enclosing.fileLinkers;
     this.protoTypeNames = enclosing.protoTypeNames;
+    this.publicImports = enclosing.publicImports;
     this.imports = enclosing.imports;
     this.contextStack = Util.concatenate(enclosing.contextStack, additionalContext);
     this.errors = enclosing.errors;
   }
 
-  public Schema link() {
-    // Register the types.
-    for (ProtoFile protoFile : protoFiles) {
-      for (Type type : protoFile.types()) {
-        register(type);
-      }
+  // TODO(jwilson): replace Iterables with an interface that loads path files on-demand.
+  public Schema link(Iterable<ProtoFile> sourceFiles, Iterable<ProtoFile> pathFiles) {
+    for (ProtoFile protoFile : sourceFiles) {
+      fileLinkers.put(protoFile.location().getPath(),
+          new FileLinker(protoFile, withContext(protoFile)));
+    }
+    for (ProtoFile protoFile : pathFiles) {
+      fileLinkers.put(protoFile.location().getPath(),
+          new FileLinker(protoFile, withContext(protoFile)));
     }
 
-    // Link extensions. This depends on type registration.
-    for (ProtoFile protoFile : protoFiles) {
-      Linker linker = withContext(protoFile);
-      for (Extend extend : protoFile.extendList()) {
-        extend.link(linker);
-      }
+    for (FileLinker fileLinker : fileLinkers.values()) {
```

Comment:
Unlikely-to-be-observed behavior change here: I’m doing imports early now. Previously we did ’em late. It’s necessary to do ’em early if we’re going to defer loading everything.